File: frontend/src/utils/cacheManager.js
import Redis from 'redis';

/**
 * Redis caching utility with write-through strategy
 */
class CacheManager {
  constructor() {
    this.client = null;
    this.isConnected = false;
    this.defaultTTL = 3600; // 1 hour default
    this.embeddingTTL = 86400; // 24 hours for embeddings
    this.responseTTL = 1800; // 30 minutes for API responses
  }

  /**
   * Initialize Redis connection
   */
  async connect() {
    if (this.isConnected) return;

    try {
      // Try to connect to Redis (local or cloud)
      const redisUrl = process.env.REDIS_URL || 'redis://localhost:6379';
      
      this.client = Redis.createClient({
        url: redisUrl,
        socket: {
          connectTimeout: 5000,
          lazyConnect: true
        },
        retry_strategy: (options) => {
          if (options.error && options.error.code === 'ECONNREFUSED') {
            console.log('Redis connection refused - running without cache');
            return false;
          }
          if (options.total_retry_time > 1000 * 60 * 60) {
            return false;
          }
          return Math.min(options.attempt * 100, 3000);
        }
      });

      await this.client.connect();
      this.isConnected = true;
      console.log('Redis connected successfully');
    } catch (error) {
      console.log('Redis connection failed - running without cache:', error.message);
      this.client = null;
      this.isConnected = false;
    }
  }

  /**
   * Get value from cache
   * @param {string} key - Cache key
   * @returns {Promise<any|null>} Cached value or null
   */
  async get(key) {
    if (!this.isConnected || !this.client) return null;

    try {
      const value = await this.client.get(key);
      return value ? JSON.parse(value) : null;
    } catch (error) {
      console.error('Cache get error:', error);
      return null;
    }
  }

  /**
   * Set value in cache with TTL
   * @param {string} key - Cache key
   * @param {any} value - Value to cache
   * @param {number} ttl - Time to live in seconds
   */
  async set(key, value, ttl = this.defaultTTL) {
    if (!this.isConnected || !this.client) return;

    try {
      await this.client.setEx(key, ttl, JSON.stringify(value));
    } catch (error) {
      console.error('Cache set error:', error);
    }
  }

  /**
   * Cache embedding with write-through strategy
   * @param {string} text - Text that was embedded
   * @param {Array} embedding - The embedding vector
   */
  async cacheEmbedding(text, embedding) {
    const key = `embedding:${this.hashText(text)}`;
    await this.set(key, embedding, this.embeddingTTL);
  }

  /**
   * Get cached embedding
   * @param {string} text - Text to get embedding for
   * @returns {Promise<Array|null>} Cached embedding or null
   */
  async getCachedEmbedding(text) {
    const key = `embedding:${this.hashText(text)}`;
    return await this.get(key);
  }

  /**
   * Cache API response
   * @param {string} question - User question
   * @param {string} response - AI response
   * @param {Object} metadata - Additional metadata
   */
  async cacheResponse(question, response, metadata = {}) {
    const key = `response:${this.hashText(question.toLowerCase())}`;
    const cacheData = {
      response,
      metadata,
      timestamp: new Date().toISOString()
    };
    await this.set(key, cacheData, this.responseTTL);
  }

  /**
   * Get cached response
   * @param {string} question - User question
   * @returns {Promise<Object|null>} Cached response data or null
   */
  async getCachedResponse(question) {
    const key = `response:${this.hashText(question.toLowerCase())}`;
    return await this.get(key);
  }

  /**
   * Cache session data
   * @param {string} sessionId - Session identifier
   * @param {Array} messages - Conversation messages
   */
  async cacheSession(sessionId, messages) {
    const key = `session:${sessionId}`;
    await this.set(key, { messages, lastUpdated: Date.now() }, 7200); // 2 hours
  }

  /**
   * Get session data
   * @param {string} sessionId - Session identifier
   * @returns {Promise<Object|null>} Session data or null
   */
  async getSession(sessionId) {
    const key = `session:${sessionId}`;
    return await this.get(key);
  }

  /**
   * Cache vector search results
   * @param {string} query - Search query
   * @param {Array} results - Search results
   */
  async cacheSearchResults(query, results) {
    const key = `search:${this.hashText(query)}`;
    await this.set(key, results, 1800); // 30 minutes
  }

  /**
   * Get cached search results
   * @param {string} query - Search query
   * @returns {Promise<Array|null>} Cached results or null
   */
  async getCachedSearchResults(query) {
    const key = `search:${this.hashText(query)}`;
    return await this.get(key);
  }

  /**
   * Increment rate limit counter
   * @param {string} identifier - User/IP identifier
   * @param {number} windowSeconds - Time window in seconds
   * @returns {Promise<number>} Current count
   */
  async incrementRateLimit(identifier, windowSeconds = 3600) {
    if (!this.isConnected || !this.client) return 0;

    try {
      const key = `ratelimit:${identifier}`;
      const current = await this.client.incr(key);
      
      if (current === 1) {
        await this.client.expire(key, windowSeconds);
      }
      
      return current;
    } catch (error) {
      console.error('Rate limit error:', error);
      return 0;
    }
  }

  /**
   * Clear cache by pattern
   * @param {string} pattern - Redis key pattern
   */
  async clearPattern(pattern) {
    if (!this.isConnected || !this.client) return;

    try {
      const keys = await this.client.keys(pattern);
      if (keys.length > 0) {
        await this.client.del(keys);
      }
    } catch (error) {
      console.error('Cache clear error:', error);
    }
  }

  /**
   * Hash text for consistent cache keys
   * @param {string} text - Text to hash
   * @returns {string} Hash string
   */
  hashText(text) {
    // Simple hash function for cache keys
    let hash = 0;
    for (let i = 0; i < text.length; i++) {
      const char = text.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(36);
  }

  /**
   * Health check for cache
   * @returns {Promise<boolean>} Cache health status
   */
  async healthCheck() {
    if (!this.isConnected || !this.client) return false;

    try {
      await this.client.ping();
      return true;
    } catch (error) {
      console.error('Cache health check failed:', error);
      this.isConnected = false;
      return false;
    }
  }

  /**
   * Close Redis connection
   */
  async disconnect() {
    if (this.client && this.isConnected) {
      await this.client.quit();
      this.isConnected = false;
    }
  }
}

// Singleton instance
const cacheManager = new CacheManager();

export default cacheManager;

File: backend/app/utils/cache_manager.py
import redis
import json
import os
from typing import Optional, Dict, Any
import time

class CacheManager:
    def __init__(self):
        self.redis_client = None
        self.connected = False
    
    async def connect(self):
        """Connect to Redis with enhanced error handling"""
        if self.connected:
            return
        
        try:
            redis_url = os.getenv("REDIS_URL")
            print(f"Attempting Redis connection with URL: {redis_url[:50]}..." if redis_url else "No REDIS_URL found")
            
            if redis_url:
                # Enhanced Redis connection with timeout and retry settings
                self.redis_client = redis.from_url(
                    redis_url, 
                    decode_responses=True,
                    socket_timeout=10,
                    socket_connect_timeout=10,
                    retry_on_timeout=True,
                    health_check_interval=30
                )
                # Test connection
                ping_result = self.redis_client.ping()
                self.connected = True
                print(f"✅ Connected to Redis successfully - Ping: {ping_result}")
                
                # Test basic operations
                self.redis_client.set("test:connection", "success", ex=60)
                test_result = self.redis_client.get("test:connection")
                print(f"✅ Redis operations test: {test_result}")
                
            else:
                print("❌ No Redis URL provided in environment variables")
                self.redis_client = None
        except redis.AuthenticationError as error:
            print(f"❌ Redis authentication error: {error}")
            self.redis_client = None
        except redis.ConnectionError as error:
            print(f"❌ Redis connection error: {error}")
            self.redis_client = None
        except Exception as error:
            print(f"❌ Redis connection failed: {error}")
            self.redis_client = None
    
    async def get_session(self, session_id: str) -> Optional[Dict[str, Any]]:
        """Get session data"""
        try:
            if self.redis_client:
                data = self.redis_client.get(f"session:{session_id}")
                return json.loads(data) if data else None
            return None
        except Exception as error:
            print(f"Error getting session: {error}")
            return None
    
    async def cache_session(self, session_id: str, session_data: Dict[str, Any]):
        """Cache complete session data including entities"""
        try:
            if self.redis_client:
                # Update timestamp
                session_data["lastUpdated"] = time.time()
                self.redis_client.setex(
                    f"session:{session_id}", 
                    3600,  # 1 hour expiry
                    json.dumps(session_data)
                )
        except Exception as error:
            print(f"Error caching session: {error}")
    
    async def get_cached_response(self, cache_key: str) -> Optional[Dict[str, Any]]:
        """Get cached response for a cache key"""
        try:
            if self.redis_client:
                cache_redis_key = f"response:{hash(cache_key)}"
                data = self.redis_client.get(cache_redis_key)
                return json.loads(data) if data else None
            return None
        except Exception as error:
            print(f"Error getting cached response: {error}")
            return None

    async def cache_response(self, cache_key: str, response: str, metadata: Dict[str, Any]):
        """Cache a response"""
        try:
            if self.redis_client:
                cache_redis_key = f"response:{hash(cache_key)}"
                cache_data = {
                    "response": response,
                    "timestamp": time.time(),
                    "metadata": metadata
                }
                self.redis_client.setex(
                    cache_redis_key,
                    1800,  # 30 minutes expiry
                    json.dumps(cache_data)
                )
        except Exception as error:
            print(f"Error caching response: {error}")
